(1349) Assessor can make state changes to application

richpjames opened the pull request:
# Context
An assessment needs to be able to move between the states outlined in this diagram (with the adjustment that from the rejected state we can only transition to unallocated).
![image](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/44123869/1bdf6659-823f-482e-a24d-d93591a4a9e0)

In order to do this we need to:
- be able to 'show' an assessment
- be able to update the state to one of the viable states, given the state we're in 
- confirm this transition
- perform it
- show the user confirmation that the assessment has been updated

This work builds on the the `feature/1346-an-assessor-can-see-an-application-listing-page-from-the-dashboard` branch which should be merged first

[Trello card](https://trello.com/c/ZI5W46gr/1349-an-assessor-can-make-state-changes-to-an-application)

# Changes in this PR
- We update the Assessments used throughout the application to the correct domain: `TemporaryAccommodationAssessment`
- We add a temporary type for the `TemporaryAccommodationAssessment` with the desired statuses as this has not yet been added in the [api spec](https://github.com/ministryofjustice/hmpps-approved-premises-api/blob/main/src/main/resources/static/api.yml). This will need to be deleted once the spec is correct
- We add the 'find' method to the assessment service
- We add some utils for generating the 'status' badge and 'actions' menu on the assessments/show view
- We add the 'show' view 
- We add the confirm view
- We add the client, service and controller methods to update the assessment
- We extend the intergration tests

## Screenshots of UI changes

![Apply -- Assess -- I can change the state of the referral from unallocated to in review (2)](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/44123869/5402a1f9-6283-4a24-8602-2afbe9bd5446)
![confirm-in-review (11)](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/44123869/244c9848-60ef-44bd-aee6-4878cca1d2e3)
![confirm-ready-to-place](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/44123869/67d8ae58-7940-46c8-a06a-381a6f406a09)
![confirm-close](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/44123869/163c7549-495a-41cf-9f7e-bc156f81dc45)
![show-with-banner](https://github.com/ministryofjustice/hmpps-temporary-accommodation-ui/assets/44123869/a7af5d81-e568-4d7f-8183-c59f65257947)





# Release checklist

[Release process documentation](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/edit-v2/4247847062?draftShareId=a1c360ab-bd31-4db1-aae3-7cc002761de9).

## Pre-merge checklist

- [ ] Are any changes required to dependent services for this change to work?
  (eg. CAS API)
- [ ] Have they been released to production already?

## Post-merge checklist

- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to test
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to preprod
- [ ] [Manually approve](https://dsdmoj.atlassian.net/wiki/spaces/AP/pages/4247847062/Release+process#Manual-releases)
  release to prod

<!-- Should a release fail at any step, you as the author should now lead the work to
fix it as soon as possible. You can monitor deployment failures in CircleCI
itself and application errors are found in
[Sentry](https://ministryofjustice.sentry.io/issues/?project=4504129156218880&referrer=sidebar&statsPeriod=24h).
Both events should be automatically sent to our [Slack
channel](https://mojdt.slack.com/archives/C048BJS7S2F). -->

Commit:
Add 'show' method to Assessments Controller

This will be used to render the view of the assessment.
I have opted to have the actions for the assessment and the status badge
mapped here, following from the BookingController's precedent but
it may be more appropriate to do this in the view or the even the service.
I add the route in the manage routes file as this is where the other assessment
route lived but I think we may want an 'assess' specific routes file in time

Commit:
Add 'find' method to Assessment service

We also add the mock API endpoint for finding an assessment

Commit:
Update assessment factory to return TA Assessments

Previously this returned AP Assessments, the status field will differ
soon so lets update all the places that assessments are used.
I also used this opportunity to alias TemporaryAccommodationAssessment
to Assessment in some more places.

Commit:
Add 'show' view and test

Here we add the view. In time it will display the content of the assessment
but for now it just displays the person's name, the assessment's status
and the list of actions that can be taken given that assessment's status.

We also test that all the content we expect to be there is

Commit:
Add 'confirm' method to Assessments controller

We have different copy here based on the status so we add an object
to select the correct one.
We also pass the status and assessment id through to the view
to enable the PUT request to be made.

Commit:
Add 'confirm' view

Commit:
Add state transition methods to Assessments client

Once the user has 'confirmed' the state change we want to update the status of the assessment in the API so we need a client methods to do that. We also add the necessary stubs so we can test these methods locally

Commit:
Add 'update' method to Assessments service

Using the provided assessment state, we call the appropriate client method

Commit:
Add 'update' method to Assessments Controller

Here we pull the status from the query params and update the assessment with that. Once the action is successful (and we assume it always will be...) we redirect to the 'show' page. We also extend the tests here to go through to the show page again

Commit:
Add further integration tests

Now we have the journey complete we can go through all the way from 'unallocated' to 'closed'

Commit:
Add assessmentActions

bryangaledxw commented on server/utils/assessmentsUtils.ts:
```diff
@@ -74,3 +78,61 @@ const dateValue = (date: string) => {
+    close: {
+      text: 'Close',
+      classes: 'govuk-button--secondary',
+      href: paths.assessments.confirm({ id: assessment.id, status: 'closed' }),
```
Suggest add paths in their own commit ahead of this

bryangaledxw commented on server/views/temporary-accommodation/assessments/show.njk:
```diff
@@ -0,0 +1,26 @@
+  {% include "../../_messages.njk" %}
+
+  {{
+        mojIdentityBar({
```
Please indent with two spaces

bryangaledxw commented on server/controllers/temporary-accommodation/manage/assessmentsController.ts:
```diff
@@ -40,6 +40,33 @@ export const archivedAssessmentsTableHeaders = [
+      <p class="govuk-body">Some regions will require approval before a bedspace is booked. Other regions require an address before approvals can be given. Check with your manager if you are unsure.</p>
+      <p class="govuk-body">Once a person has been placed the referral should be archived.</p>`,
+  },
+  rejection: {
```
Should this be "rejected" rather than "rejection", to match the statuses on the type?

bryangaledxw commented on server/controllers/temporary-accommodation/manage/assessmentsController.ts:
```diff
@@ -40,6 +40,33 @@ export const archivedAssessmentsTableHeaders = [
     },
   },
 ]
+export const confirmationPageContent: Record<string, Record<'title' | 'text', string>> = {
```
Can the type be more precise? Does:

`Record<Assessment['status'], {title: string, text: string}>`

work?

bryangaledxw commented on server/views/temporary-accommodation/assessments/confirm.njk:
```diff
@@ -0,0 +1,22 @@
+    {{ govukButton({
+          text: "Confirm",
+          preventDoubleClick: true
+        }) }}
```
Please fix indentation

bryangaledxw reviewed (commented):
Will do another pass once the underlying PR is in, but I've made some interim comments

Commit:
Add path for confirmation screen

This will be used for showing the confirmation screens in a subsequent commit.
We call the path function with a query param for the status so we can
work out the correct confirmation message to show

Commit:
Add describe block for list test

bryangaledxw commented on server/services/assessmentsService.test.ts:
```diff
@@ -97,4 +97,18 @@ describe('AssessmentsService', () => {
+
+      assessmentClient.find.mockResolvedValue(assessment)
+
+      expect(service.findAssessment(callConfig, assessmentId)).resolves.toEqual(assessment)
```
I *think* this wants an `await`

bryangaledxw commented on server/controllers/temporary-accommodation/manage/assessmentsController.ts:
```diff
@@ -59,4 +60,18 @@ export default class AssessmentsController {
+      return res.render('temporary-accommodation/assessments/show', {
+        assessment,
+        actions: assessmentActions(assessment),
+        statusBadge: statusTag(assessment.status),
```
I realise we're super inconsistent in this, which isn't helpful 😅. My feeling is that status tag is a small enough, purely presentation enough, thing that it's better to call in the njk file than the controller. My suggestion is rename the existing nunjucks `statusTag` to `personStatusTag` and expose this as 'assessmentStatusTag', rather than adding a new way of getting the statusTag into the view

bryangaledxw commented on integration_tests/tests/temporary-accommodation/assess/assess.cy.ts:
```diff
@@ -16,47 +16,49 @@ context('Apply', () => {
   })
 
   describe('Assess', () => {
-    it('I can view a list of assessments', () => {
```
Thanks for doing this in it's own commit 👍

bryangaledxw commented on server/views/temporary-accommodation/assessments/show.njk:
```diff
@@ -0,0 +1,26 @@
+
+{% endblock %}
+
+{% block extraScripts %}
+  <script type="text/javascript" nonce="{{ cspNonce }}">
+    new MOJFrontend.ButtonMenu({container: $('.moj-button-menu'), mq: "(min-width: 200em)", buttonText: "Update referral status", menuClasses: "moj-button-menu__wrapper--right"});
+  </script>
+{% endblock %}
```
Suggest wrap this in `{% if actions %}` same as elsewhere - it's caught me off-guard before that this function causes browser-side JS errors if actions is null

bryangaledxw commented on server/controllers/temporary-accommodation/manage/assessmentsController.ts:
```diff
@@ -30,6 +31,33 @@ export const assessmentsTableHeaders = [
   },
 ]
 
+export const confirmationPageContent: Record<Assessment['status'], Record<'title' | 'text', string>> = {
```
Few instances of ` quotes that could be plain ' quotes in this block

bryangaledxw replied:
And I think this type can still be more precise - {title: string, text: string} says they both want to be there, and I think is more straightforward than Record<'title' | 'text', string>. And I expect there's a way to enforce that this covers all statuses (do we want this to cause a compile error if new statuses are added we're not handling here?)

richpjames replied:
I've used backticks for multiline strings as I think it's more readable than concatenating the lines or having very long line, happy to change it if you disagree

bryangaledxw replied:
No, that's totally fair!

bryangaledxw commented on server/routes/temporary-accommodation/manage.ts:
```diff
@@ -304,6 +304,9 @@ export default function routes(controllers: Controllers, services: Services, rou
     auditEvent: 'VIEW_ASSESSMENT',
   })
+  get(paths.assessments.confirm.pattern, assessmentsController.confirm(), {
+    auditEvent: 'CONFIRM_ASSESMENT_STATUS_CHANGE',
```
I think this wants to be something like `VIEW_ASSESSMENT_STATUS_CHANGE_CONFIRM` - generally our `get`s are VIEW_s

bryangaledxw commented on server/controllers/temporary-accommodation/manage/assessmentsController.ts:
```diff
@@ -112,4 +114,18 @@ export default class AssessmentsController {
+      assessment.status = req.params.status as Assessment['status']
+      await this.assessmentsService.updateAssessment(callConfig, assessment)
+
+      req.flash('info', `Assessment updated status updated to "${lowerCase(assessment.status)}"`)
```
I'm a bit wary about tying user-facing naming to API naming - thought on `lowerCase(statusName(assessment.status))`?

richpjames replied:
yeah makes sense

bryangaledxw commented on server/routes/temporary-accommodation/manage.ts:
```diff
@@ -308,5 +308,9 @@ export default function routes(controllers: Controllers, services: Services, rou
 
+  put(paths.assessments.update.pattern, assessmentsController.update(), {
+    auditEvent: 'UPDATE_ASSESSMENT',
+  })
```
My feeling is use the same conventions as other `puts`, and use `redirectAuditEventSpecs`, even though we only have the one possible redirect destination

bryangaledxw reviewed (approved):
Lovely stuff! Few minor comments, but otherwise LGTM

Commit:
Rename statusTag to personStatusTag

We will soon add an assessmentStatusTag so this keeps things clear

bryangaledxw committed:
Add assertUnreachable() helper function

We add an assertUnreachable() function that we can use to ensure at compile time that a switch statement checks all possible cases for a given input

bryangaledxw committed:
Move retrieveQuestionResponseFromApplication()

We move the function from utils to applicationUtils. We will want to use assertUnreachable() in our RestClient class, and without moving this function, we would create a circular dependency

Update applicationUtils.test.ts

bryangaledxw committed:
Add delete() method to ResetClient

To move assessments through their possible states, we will want to make DELETE calls to the API, so we expand postOrPut() to postPutOrDelete()

bryangaledxw committed:
Add breadcrumbs to assessment listing page

bryangaledxw commented:
@tahb Do you want to take a look at this before I merge, so I'm not approving my own changes?

Most of my changes are the first 4 commits, plus "Add state transition methods to Assessments client", and "Add 'update' method to Assessments service"

patinipat reviewed (approved):
code LGTM
still need to setup my backend to test drive though
